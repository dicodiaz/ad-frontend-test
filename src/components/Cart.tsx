'use client';

import { useCartContext } from '@/context';
import { CartItem } from '.';

export const Cart: React.FC = () => {
  const { cart } = useCartContext();

  return (
    <section className="flex flex-col gap-y-8 px-6 py-8">
      <div className="flex flex-col gap-y-3 text-[#3B3B3B]">
        <h1 className="text-2xl font-bold leading-7 tracking-wide">Your cart</h1>
        <span className="text-xl leading-6 tracking-wide">
          {cart.length} {cart.length === 1 ? 'item' : 'items'}
        </span>
      </div>
      <div className="flex flex-col">
        {cart.map((game) => (
          <CartItem key={game.id} game={game} />
        ))}
      </div>
    </section>
  );
};
