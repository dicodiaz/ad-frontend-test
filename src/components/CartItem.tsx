import { useCartContext } from '@/context';
import { Game } from '@/utils/endpoint';
import { HiX } from 'react-icons/hi';

export type CartItemProps = {
  game: Game;
};

export const CartItem: React.FC<CartItemProps> = ({ game }) => {
  const { id, image, genre, name, price, description, isNew } = game;
  const { dispatch } = useCartContext();

  const removeFromCart = () => {
    dispatch({ type: 'REMOVE_FROM_CART', payload: id });
  };

  return (
    <div className="flex flex-col gap-y-4 px-4 py-5 [&~&]:border-t-[0.5px] [&~&]:border-[#8F8F8F]">
      <div className="relative flex gap-x-3">
        {isNew && (
          <div className="absolute left-3 top-3 rounded bg-[#F5F5F4] px-3 py-2 leading-4 tracking-wide">
            New
          </div>
        )}
        <img src={image} alt={name} className="h-[135px] flex-1 object-cover" />
        <div>
          <button type="button" className="p-1" onClick={removeFromCart}>
            <HiX className="size-5 text-[#8F8F8F]" />
          </button>
        </div>
      </div>
      <div className="flex flex-col gap-y-11">
        <div className="flex flex-col gap-y-3">
          <span className="text-sm font-bold leading-4 text-[#737373]">{genre}</span>
          <div className="flex flex-col gap-y-2">
            <h4 className="text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">{name}</h4>
            <span className="leading-5 text-[#737373]">{description}</span>
          </div>
        </div>
        <div className="flex justify-end text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">
          {price.toLocaleString('en-US', { style: 'currency', currency: 'USD' })}
        </div>
      </div>
    </div>
  );
};
