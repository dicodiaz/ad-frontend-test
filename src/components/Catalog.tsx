import { getDataFromServer } from '@/services';
import { GameCard, GenreSelect, SeeMoreButton } from '.';

export type CatalogProps = {
  genre?: string;
  currentPage: number;
};

export const Catalog: React.FC<CatalogProps> = async ({ genre, currentPage }) => {
  const { games, availableFilters, totalPages } = await getDataFromServer({
    genre,
    page: currentPage,
  });

  return (
    <section className="flex flex-col gap-y-2">
      <div className="flex flex-col gap-y-8">
        <GenreSelect availableFilters={availableFilters} />
        <div className="h-px bg-[#EFEDF3]" />
        <div className="flex flex-col gap-y-6 px-6">
          {games.map((game) => (
            <GameCard key={game.id} game={game} />
          ))}
        </div>
      </div>
      {currentPage < totalPages && <SeeMoreButton currentPage={currentPage} />}
    </section>
  );
};
