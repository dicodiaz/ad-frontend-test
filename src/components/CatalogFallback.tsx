import { ImSpinner2 } from 'react-icons/im';

export const CatalogFallback: React.FC = () => {
  return (
    <div className="flex justify-center px-6">
      <ImSpinner2 className="size-6 animate-spin" />
    </div>
  );
};
