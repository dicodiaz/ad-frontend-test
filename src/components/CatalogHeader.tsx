'use client';

import { Data } from '@/services';
import { useRouter, useSearchParams } from 'next/navigation';

export type CatalogHeaderProps = {
  availableFilters: Data['availableFilters'];
};

export const CatalogHeader: React.FC<CatalogHeaderProps> = ({ availableFilters }) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const handleGenreChange: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    const genre = e.target.value;
    const params = new URLSearchParams(searchParams);
    if (genre === 'all') {
      params.delete('genre');
    } else {
      params.set('genre', genre);
    }
    router.push(`?${params.toString()}`);
  };

  return (
    <header className="flex flex-col gap-y-8 border-b border-[#EFEDF3] px-6 py-8">
      <h1 className="text-2xl font-bold leading-7 tracking-[0.4px] text-[#3B3B3B]">TOP SELLERS</h1>
      <label className="flex gap-x-6 py-4">
        <span>Genre</span>
        <div className="my-[1px] w-px bg-[#3B3B3B]" />
        <select className="flex-1" onChange={handleGenreChange}>
          <option value="all">All</option>
          {availableFilters
            .sort((a, b) => a.localeCompare(b))
            .map((genre, index) => (
              <option key={index} value={genre}>
                {genre}
              </option>
            ))}
        </select>
      </label>
    </header>
  );
};
