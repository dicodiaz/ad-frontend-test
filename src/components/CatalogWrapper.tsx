import { getDataFromServer } from '@/services';
import { Catalog } from '.';

export type CatalogWrapperProps = {
  genre?: string;
};

export const CatalogWrapper: React.FC<CatalogWrapperProps> = async ({ genre }) => {
  const data = await getDataFromServer({ genre });

  return <Catalog data={data} />;
};
