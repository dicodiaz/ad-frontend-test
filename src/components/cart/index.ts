export * from './Cart';
export * from './CartItem';
export * from './OrderSummary';
