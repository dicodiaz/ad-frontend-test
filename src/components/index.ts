export * from './cart';
export * from './checkout';
export * from './home';
export * from './layout';
