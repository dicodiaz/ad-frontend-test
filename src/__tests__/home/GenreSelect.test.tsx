import { GenreSelect } from '@/components';
import { availableFilters } from '@/utils/endpoint';
import { fireEvent, render, screen } from '@testing-library/react';

describe('GenreSelect', () => {
  beforeEach(() => {
    (globalThis as any).mockPush.mockClear();
    (globalThis as any).mockSearchParams = new URLSearchParams(); // reset
  });

  it('renders GenreSelect unchanged', () => {
    const { container } = render(<GenreSelect availableFilters={availableFilters} />);
    expect(container).toMatchSnapshot();
  });

  it('defaults to "all" if no genre in searchParams', () => {
    render(<GenreSelect availableFilters={availableFilters} />);
    expect(screen.getByRole('combobox')).toHaveValue('all');
  });

  it('shows selected genre if genre is in searchParams', () => {
    (globalThis as any).mockSearchParams = new URLSearchParams('genre=Action');
    render(<GenreSelect availableFilters={availableFilters} />);
    expect(screen.getByRole('combobox')).toHaveValue('Action');
  });

  it('updates URL when genre changes', () => {
    render(<GenreSelect availableFilters={availableFilters} />);
    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'Horror' } });

    expect((globalThis as any).mockPush).toHaveBeenCalledWith('?genre=Horror');
  });

  it('removes genre param if "all" is selected', () => {
    (globalThis as any).mockSearchParams = new URLSearchParams('genre=Action');
    render(<GenreSelect availableFilters={availableFilters} />);
    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'all' } });

    expect((globalThis as any).mockPush).toHaveBeenCalledWith('?');
  });
});
