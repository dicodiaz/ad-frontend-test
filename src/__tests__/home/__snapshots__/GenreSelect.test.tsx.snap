// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GenreSelect renders GenreSelect unchanged 1`] = `
<div>
  <select
    aria-label="Select genre"
    class="flex-grow md:flex-grow-0"
  >
    <option
      selected=""
      value="all"
    >
      All
    </option>
    <option
      value="Action"
    >
      Action
    </option>
    <option
      value="Action-Adventure"
    >
      Action-Adventure
    </option>
    <option
      value="Adventure"
    >
      Adventure
    </option>
    <option
      value="Battle Royale"
    >
      Battle Royale
    </option>
    <option
      value="Horror"
    >
      Horror
    </option>
    <option
      value="Indie"
    >
      Indie
    </option>
    <option
      value="MMORPG"
    >
      MMORPG
    </option>
    <option
      value="MOBA"
    >
      MOBA
    </option>
    <option
      value="Puzzle"
    >
      Puzzle
    </option>
    <option
      value="Racing"
    >
      Racing
    </option>
    <option
      value="RPG"
    >
      RPG
    </option>
    <option
      value="Shooter"
    >
      Shooter
    </option>
    <option
      value="Simulation"
    >
      Simulation
    </option>
    <option
      value="Sports"
    >
      Sports
    </option>
    <option
      value="Strategy"
    >
      Strategy
    </option>
  </select>
</div>
`;
